### PR TITLE
Feat issue 137

### PR DESCRIPTION
**Pull request by laurent-bientz**
Refering to #137, the purpose of this merge request is to relocate the process of the Twig Extension `label()` method in an external service to allow using it outside of a Twig rendering (for my case, it was for displaying the "human" value behind the `Enum` in flash messages).

What I did:
- created a new service
- loaded the new services (services.xml)
- tweaked the construct of Twig Extension to inject the new service instead of the Translator
- tweaked the CompilerPass to inject the Translator in the service instead of the Twig Extension
- updated tests as well

Please note that after registering the bundle, I had to alias the new service to inject it on my Controller:
```yaml
services:
    Greg0ire\Enum\Bridge\Symfony\Translator\Label: "@greg0ire_enum.symfony.translator.label"
```

Not really sure what happened, I tried to debug autowiring but didn't understand, maybe a problem of FQN.

Also, I was not really sure how to name the new Service and where place it ;) so I tried to follow your structure and naming conventions, feel free to provide a better name/tree.

Thanks!

### CODE REVIEW

**Comment by laurent-bientz** [comment]
About the `OSKAR-phpstan` job, it seems to be related to Symfony autoloading

**Comment by greg0ire** [comment]
Currently working on it :  #139 :)

File: src/Bridge/Symfony/Translator/Label.php [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+    /**
+     * @var TranslatorInterface
+     */
+    private $translator;
+
+    /**
+     * Label constructor.
+     *
+     * @param TranslatorInterface $translator
+     */
```
Comment (greg0ire) [inline src/Bridge/Symfony/Translator/Label.php]: Let's remove that phpdoc please
```suggestion
```

File: src/Bridge/Twig/Extension/EnumExtension.php [hunk 2]
```diff
@@ -14,16 +14,13 @@
...
+     * @var Label
      */
-    private $translator;
+    private $label;
 
-    /**
-     * @param TranslatorInterface $translator
-     */
-    public function __construct(TranslatorInterface $translator = null)
+    public function __construct(Label $label)
```
Comment (greg0ire) [inline src/Bridge/Twig/Extension/EnumExtension.php]: This is a BC-break. Let's fix this:
```suggestion
    public function __construct($label)
    {
        if ($label instanceof TranslatorInterface) {
            @trigger_error(sprintf(
                'Providing a %s instance to %s is deprecated and will not be supported in 5.0. Please provide a %s instance instead.'
                TranslatorInterface::class,
                __METHOD__,
                Label::class
            ), E_USER_DEPRECATED);
            $this->label = new Label($label);
            
            return;
        }
        $this->label = $label;
    }
```

File: src/Bridge/Symfony/Translator/Label.php [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+     * @param string|bool $translationDomain  the translation domain to use if the translator if available.
+     *                                        string: Use the specified one
+     *                                        null: Use the default one
+     *                                        false: Do not use the translator
+     * @param bool        $classPrefixed      Prefix the label with the enum class. Defaults to true if the translator
+     *                                        is available and enabled, false otherwise.
+     * @param string      $namespaceSeparator namespace separator to use with the class prefix.
+     *                                        This takes effect only if $classPrefixed is true
+     */
+    public function run(
```
Comment (greg0ire) [inline src/Bridge/Symfony/Translator/Label.php]:
```suggestion
    public function __invoke(
```

File: src/Bridge/Twig/Extension/EnumExtension.php [hunk 4]
```diff
@@ -67,27 +64,7 @@ public function label(
...
-        );
-
-        if ($useTranslation) {
-            $translatedLabel = $this->translator->trans($label, [], $translationDomain);
-
-            return $translatedLabel ?: $label;
-        }
-
-        return $label;
+        return $this->label->run($value, $class, $translationDomain, $classPrefixed, $namespaceSeparator);
```
Comment (greg0ire) [inline src/Bridge/Twig/Extension/EnumExtension.php]:
```suggestion
        return ($this->label)($value, $class, $translationDomain, $classPrefixed, $namespaceSeparator);
```

File: src/Bridge/Symfony/Translator/Label.php [hunk 5]
```diff
@@ -0,0 +1,67 @@
+<?php
+
+namespace Greg0ire\Enum\Bridge\Symfony\Translator;
+
+use Symfony\Contracts\Translation\TranslatorInterface;
+
+final class Label
```
Comment (greg0ire) [inline src/Bridge/Symfony/Translator/Label.php]: Let's name it as an action and make it invokable
```suggestion
final class GetLabel
```

**Comment by greg0ire** [comment]
> Not really sure what happened, I tried to debug autowiring but didn't understand, maybe a problem of FQN.

I think this is expected. Maybe it could work without the alias on old versions of Symfony, but things have gotten stricter.

**Comment by laurent-bientz** [comment]
> 
> 
> > Not really sure what happened, I tried to debug autowiring but didn't understand, maybe a problem of FQN.
> 
> I think this is expected. Maybe it could work without the alias on old versions of Symfony, but things have gotten stricter.

Not a big deal btw, my project was running Sf 4.4.

I finished the review, good idea for the __invoke.

Just, for my culture, why we should remove some phpDoc? it's because of the TypeHinting which is enough? I saw in the FlintCI that I had to remove `@return` annotation

**Comment by greg0ire** [comment]
> Just, for my culture, why we should remove some phpDoc? it's because of the TypeHinting which is enough?

Yes exactly. Comments are supposed to be help for the code. They have been hijacked for phpdoc as help for the IDE and SA tools, but now that we have type declarations, we humans should reclaim the comments back! Let's have comments directed at humans :)

**Comment by laurent-bientz** [comment]
> 
> 
> > Just, for my culture, why we should remove some phpDoc? it's because of the TypeHinting which is enough?
> 
> Yes exactly. Comments are supposed to be help for the code. They have been hijacked for phpdoc as help for the IDE and SA tools, but now that we have type declarations, we humans should reclaim the comments back! Let's have comments directed at humans :)

Crystal clear!

Let me review some things, I accepted to quickly your suggests ;) some code is broken

**Comment by greg0ire** [comment]
Yes, my suggests are incomplete, you are going to have to rename the file where `GetLabel` is defined, and some other things, take your time.

**Comment by laurent-bientz** [comment]
@greg0ire I think we're good!

**Comment by greg0ire** [comment]
The code looks good. Can you please document the new service and its usage in the README?

**Comment by laurent-bientz** [comment]
After dinner, sure ;)

**Comment by greg0ire** [comment]
Enjoy your dinner!

**Comment by greg0ire** [comment]
Good job writing the documentation, it's really clear :ok_hand:

**Pull request merged**